### PR TITLE
SPP 10277 UAT failures item non response

### PR DESCRIPTION
**Pull request by gibbardsteve**
## Synopsis

With the recent release candidate UAT testing for totals and components processing was impaired with scenarios where the component list included some or all empty values.

## Checklist

- [X] Documentation created/updated
- [X] Tests created/updated

## Description

Selective Editing and Date Adjustment use Pandas NaN to imply empty cell. Totals and Components and Thousand Pounds processing has been tweak slightly to work in a similar way.

Specifically:

**thousand_pounds.py**
clean_target_variables() - where a string of "NaN" is received or returned, treat as float('Nan')
Return a string representation of numbers, but if the value is None or Nan then do not convert to string

**example.py**
The csv example is updated to ensure empty cells are treated as nan do bring inline with pandas processing

**totals_and_components.py**
clean_components_list() - where a string of "NaN" is received or returned, treat as float('Nan')

**pandas_example.py**
run_all_csvs() - allow the processing of all csvs in a given directory.  Any file that contains "output" in the name is ignored when finding a list of input data files.  This function allows all UAT files provided by MQD to be run.

**pandas_wrapper.py**
When a value in the input data is None treat as float('Nan')

**example_test_data.csv**
Add thousand pounds examples where empty cells are present for target_variables

**example_test_data.csv**
Add totals and components data to test empty cells in component list

Additional CSV files are added for totals and components and thousand pounds processing - these are the UAT files that are run by MQD

### CODE REVIEW

File: sml_small/editing/totals_and_components/example.py [hunk 1]
```diff
@@ -222,10 +223,10 @@ def invoke_process_with_local_csv():
                 (row["reference"]),
                 (row["total"]),
                 [
-                    (row["comp_1"]),
-                    (row["comp_2"]),
-                    (row["comp_3"]),
-                    (row["comp_4"]),
+                    float('Nan') if not row["comp_1"] else (row["comp_1"]),
```
Comment (sanjeevz3009) [inline sml_small/editing/totals_and_components/example.py:226]: You have not done the same checks in the _invoke_process_with_in_memory_csv_. That function does have hard-coded in-memory CSV data but it may be good to keep it consistent.

**Comment by sanjeevz3009** [comment]
When the _pandas_example.py_ runs all the UAT tests for TPC & TCC it produces around 51 output files and these aren't being ignored, unless we want these output files in the repo now. As we are outputting a lot of files now, we can remove lines 10 to 17 in the gitignore file and replace them with *_output.csv*, so it ignores anything that ends with _output.csv_ =)

If going with the above solution to ignore the output files, we also need to update the TPC _example.py_ line 110 output file name from _output_1.csv_ to _output_1.csv._

File: sml_small/utils/pandas_example.py [hunk 2]
```diff
@@ -110,9 +153,32 @@ def run_totals_components_with_pandas(path, input_csv, output_csv):
...
+    final_df = final_df[column_order]
+
+    # Write the DataFrame to a CSV, excluding the index column
+    csv_filename = path+output_csv
+    final_df.to_csv(csv_filename, index=False)
+
 
 # ---------------------------------------
 # Thousand Pounds usage with Pandas
+# Thousand Pounds usage with Pandas
```
Comment (jhmjam) [inline sml_small/utils/pandas_example.py:181]: duplicate?

File: sml_small/utils/pandas_wrapper.py [hunk 3]
```diff
@@ -132,7 +132,7 @@ def run_thousand_pounds(
     target_variables_list = {}
     for value in target_variables_columns:
         if row[value] is None:
-            target_variables_list[value] = 'Nan'
+            target_variables_list[value] = float('Nan')
```
Comment (jhmjam) [inline sml_small/utils/pandas_wrapper.py:135]: Does not really matter but should these be NaN? Not Nan

File: sml_small/editing/totals_and_components/example.py (hunk 1 above)
Comment (gibbardsteve) [reply sml_small/editing/totals_and_components/example.py]: Updated

**Comment by gibbardsteve** [comment]
> When the _pandas_example.py_ runs all the UAT tests for TPC & TCC it produces around 51 output files and these aren't being ignored, unless we want these output files in the repo now. As we are outputting a lot of files now, we can remove lines 10 to 17 in the _gitignore_ file and replace them with `*output.csv*`, so it ignores anything that ends with `output.csv` =)
> 
> If going with the above solution to ignore the output files, we also need to update the TPC _example.py_ line 110 output file name from `output1.csv` to `output.csv`.

Good point - update gitignore and changed the name of the output file.

File: sml_small/utils/pandas_example.py (hunk 2 above)
Comment (gibbardsteve) [reply sml_small/utils/pandas_example.py]: Thanks - removed

File: sml_small/utils/pandas_wrapper.py (hunk 3 above)
Comment (gibbardsteve) [reply sml_small/utils/pandas_wrapper.py]: I have made them consistent throughout the code base now, there was a mix.

**Review by jhmjam — APPROVED** [review]

**Review by sanjeevz3009 — APPROVED** [review]
Looks good to me =) Happy to approve!
